### ISSUE TITLE
Registry Token Remote Certs

### ISSUE THREAD

**Issue by FrankSpitulski** [docker/distribution#3166]
Would you all be open to accepting a PR that adds support for remote certs not bundled in the application, such as https://www.googleapis.com/oauth2/v1/certs? How about JWKS support https://www.googleapis.com/oauth2/v3/certs?

**Comment by dmcgowan** [comment]
Can you explain the use case here and what you are trying to accomplish?

**Comment by FrankSpitulski** [comment]
@dmcgowan thanks for the quick response! 

I'd like to be able to rotate the certs that the auth tokens are signed with. Since right now you have to give the registry a file to read from I'd have to redeploy the container with the new certs embedded. If the registry could read remote public keys (preferably a set like google has there) then I wouldn't need to re-package and re-deploy the registry on key change.

**Comment by dmcgowan** [comment]
Seems like it would just make sense to read from a local file. Reloading can be tricky to time correctly, but it doesn't make sense for the registry process itself to download certs. Normally in cases like this the local configuration may be handled by an external process and the registry should be able to reload configurations without bringing the whole service or container down.

**Comment by FrankSpitulski** [comment]
As best as I can tell the registry does not support reloading the local configuration.

https://github.com/docker/distribution/blob/980d1a696b5c5bed25835f25d06ec2ef3ddc24b0/registry/auth/token/accesscontroller.go#L184 is called in a `func init()`. 

Are you suggesting to watch for changes on that file and re-register the token access controller when there is a change?

**Comment by FrankSpitulski** [comment]
Digging a little deeper it seems that the access controllers are explicitly stopped from re-registering and even if that was removed the application controller is chosen during app startup so re-registering would not help.

**Comment by FrankSpitulski** [comment]
I forked and made a small change to demonstrate my solution if you'd like to take a look @dmcgowan. It also includes the remote key download option since after messing with fsnotify for a bit I thought it was much simpler this way.

https://github.com/docker/distribution/compare/master...FrankSpitulski:feat/registry/token-cert-rotation?expand=1

**Comment by FrankSpitulski** [comment]
Hi @dmcgowan, have you had a chance to check out the solution I'm going for? Of course it can be changed to better match the style of the project but it demonstrates what I had in mind.

**Comment by hobbytp** [comment]
@FrankSpitulski any update for this, this feature is useful for production environment.

**Comment by FrankSpitulski** [comment]
@hobbytp agreed, just waiting for some confirmation/guidance from the maintainers about this feature. I'd love to have it and would be happy to implement it.

**Comment by FrankSpitulski** [comment]
@dmcgowan bump

**Comment by sergioegutierrez** [comment]
Hi all, I'm also interested in contributing to the use of JWKS for production environments and I'm already working in a solution. My proposal to avoid re-registering the access controller, is to spawn a go routine that periodically update the trusted keys on the access controller based in a configurable interval (REGISTRY_AUTH_TOKEN_PULLRATE)
For this feature to be complete the following points need to be covered:

- [x] Token verification with RSA keys
- [ ] Token verification with ECDSA keys
- [x] Token verification after key rotation on IdP

I already have a change with the checked points working. Please let me know what you think about this scope and if you are ok with a PR of this work in progress to start review considerations, or if it's better to send it once its completed.

**Comment by mesaglio** [comment]
Hi all — reviving this issue with the intention of contributing a PR.

  **Context since 2020:**
  - The migration to `go-jose` (#4096) makes JWKS support more natural to implement
  - Recent auth work (#4578, #4684, #4471) shows this area is actively maintained
  - The current `rootcertbundle`-only approach still requires container redeployment to rotate signing keys

  **What I'm proposing to implement:**
  A configurable option to fetch public keys from a remote JWKS endpoint, with periodic
  background refresh (configurable interval). This would allow auth server key rotation
  without registry restarts.

  **Open questions before I start:**
  1. @dmcgowan previously preferred local file reload over remote fetch — is that still
     the maintainer preference, or is remote JWKS fetch acceptable given that go-jose
     already handles JWK parsing?
  2. Should this be additive (new config option alongside `rootcertbundle`) or a
     replacement path?
  3. Any preference on where to hook the refresh — at the token verifier level or
     the access controller level?

  Happy to open a draft PR for early feedback once the approach is agreed. @milosgajdos @thaJeztah

**Comment by milosgajdos** [comment]
Feel free to open a PR @mesaglio 

It's easier to discuss code when people can see it.
